python/ray/workflow/common.py

```diff
@@ -54,6 +62,41 @@ def __hash__(self):
         return hash(self.step_id)
 
 
+class _RefBypass:
+    """Prevents an object ref from being hooked by a serializer."""
+
+    def __init__(self, ref):
+        self._ref = ref
+
+    def __reduce__(self):
+        from ray import cloudpickle
+        return cloudpickle.loads, (cloudpickle.dumps(self._ref), )
```

Comment:
I'm not sure about this. Couldn't it cause the ref count to go to 0?

Reply:
If I'm understanding correctly, this is to avoid resolving unready workflows returned by workflow.wait, right?

How about instead setting a custom serializer for a `Workflow` created through `workflow.wait`? Then, we can serialize all the internal `Workflow`s so we don't have to decide whether to resolve a passed `Workflow` or not.

Reply:
Because we are using the ObjectRef serializer of Ray, I think the ref count etc is handled properly. This custom serializer should guarantee:

1. All `Workflow` not created by `workflow.wait` is not allowed to be serialized.
2. The object refs contained in the `Workflow` returned by `workflow.wait` should not be checkpointed, but all other refs should still be checkpointed.
3. All other objects contained in the workflow created by `workflow.wait` are serialized using the default serializer. This means a new serializer for `Workflow` may not be a good idea.

So here the simplest solution should be a custom reducer for `WorkflowStaticRef`, because the reference itself should not be serialzed, and we can contain it in other objects like the workflow or some inputs.